src/tools/pub/dependencies.md

```diff
@@ -397,23 +388,78 @@ dependency_overrides:
   transmogrify: '3.2.1'
 {% endprettify %}
 
-*Caution:* Using a dependency override involves some risk. For example,
-using an override to specify a version outside the range that the
-package claims to support, or using an override to specify
-a local copy of a package that has unexpected behaviors,
-may break your application.
+{{site.alert.warning}}
+  Using a dependency override involves some risk. For example,
+  using an override to specify a version outside the range that the
+  package claims to support, or using an override to specify
+  a local copy of a package that has unexpected behaviors,
+  may break your application.
+{{site.alert.end}}
+
+## Best practices
+
+It’s important to actively manage your dependencies and
+ensure that your packages use the freshest versions possible.
+If any dependency is stale,
+then you might have not only a stale version of that package,
+but also stale versions of other packages in your dependency graph that
+depend on that package.
+These stale versions can have a negative impact on
+the stability, performance, and quality of apps.
+
+We recommend the following best practices for package dependencies:
+
+Use [caret syntax](#caret-syntax)
+: Specifying dependencies with version ranges is such as `^1.6.3`
+  is a good practice because it allows the pub tool to
+  select newer versions of the package when they become available.
+  Also, it places an upper limit on the allowed version,
+  based on an assumption that packages use [semantic versions][],
+  where any version of path versioned `1.x` is compatible,
+  but where a new version `2.x` would be a major upgrade
+  that isn't semantically compatible with `1.x` versions. 
+
+Depend on the latest stable package versions
+: Use [`pub upgrade`][] to update to the latest package versions
+  that your pubspec allows.
+  To identify dependencies in your pubspec that
+  prevent you from updating to the latest stable versions,
+  use [`pub outdated`][].
+
+Test whenever you update package dependencies
+: If you don't update your pubspec and just run [`pub upgrade`][],
+  the API should stay the same
+  and your code should run as before — but test to make sure.
+  If you modify the pubspec and update to a new major version,
+  then you might encounter breaking changes,
+  so you need to test even more thoroughly.
+
+Don't check in the `pubspec.lock` file for library packages
+: Following the guidance in [What not to commit][],
+  don't check in `pubspec.lock` for library packages.
+  For app packages, you can commit the lock file if
+  it helps you track the dependencies that are used for
+  production or across a team.
```

Comment:
I'm not sure we should make this recommendation a _best-practice_.

Honestly, this is one of the more controversial recommendations. Some people would always want to check-in `pubspec.lock` for applications (we do that) -- otherwise dependencies may change between QA and production. Some people may even want to check-in `pubspec.lock` for libraries to get a consistent development environment.

It depends a bit on workflow, hence, I don't think it's a _best practice_.

Reply:
@mit-mit, @munificent, @kevmoo what do you think?

This section is completely new, and I kind of threw everything we've said about dependency management here... but we might want to start with less? Or to have a followup paragraph about practices that are more nuanced.

Reply:
For libraries, I think it's a 'should not', – not a 'must not'.

For libs, if you check it in, any CI environment your in will honor it, but it won't be honored for your users. I think the guidance to not check in is a great idea.

Reply:
@kevmoo, the guidance to not check in `pubspec.yaml` for libraries is reasonable. IMO a matter of personal preference: "stable CI" vs. "More relevant tests".

But for applications I think the guidance ought to strongly encourage checking `pubspec.lock` into git. Otherwise, you'll have discrepancy between devs, CI, QA and production apps.

If you don't check `pubspec.yaml` into git, you could easily ship untested code in an app store. In case a new version of a dependency is published between the time you do QA and the time you build+publish your app.